src/nupic/ntypes/Sdr.hpp

```diff
@@ -76,67 +81,79 @@ namespace nupic {
+ *    X.setFlatSparse({ 1, 4, 8 });
+ *    X.setSparse({{ 0, 1, 2,}, { 1, 1, 2 }});
  *
  *    // Access data in any format, SDR will automatically convert data formats.
  *    X.getDense()     -> { 0, 1, 0, 0, 1, 0, 0, 0, 1 }
- *    X.getIndex()     -> {{ 0, 1, 2 }, {1, 1, 2}}
- *    x.getFlatIndex() -> { 1, 4, 8 }
+ *    X.getSparse()     -> {{ 0, 1, 2 }, {1, 1, 2}}
+ *    x.getFlatSparse() -> { 1, 4, 8 }
  *
  *    // Data format conversions are cached, and when an SDR value changes the
  *    // cache is cleared.
- *    X.setFlatIndex({});  // Assign new data to the SDR, clearing the cache.
+ *    X.setFlatSparse({});  // Assign new data to the SDR, clearing the cache.
  *    X.getDense();        // This line will convert formats.
  *    X.getDense();        // This line will resuse the result of the previous line
  */
-
 class SparseDistributedRepresentation : public Serializable
 {
 private:
     vector<UInt> dimensions_;
     UInt         size_;
 
-    vector<Byte>         dense;
-    vector<UInt>         flatIndex;
-    vector<vector<UInt>> index;
+    SDR_dense_t      dense;
+    SDR_flatSparse_t flatSparse;
+    SDR_sparse_t     sparse;
 
-    bool dense_valid_;
-    bool flatIndex_valid_;
-    bool index_valid_;
+    /**
+     * These flags remember which data formats are up-to-date and which formats
+     * need to be updated.
+     */
+    bool dense_valid;
+    bool flatSparse_valid;
+    bool sparse_valid;
 
-    vector<void (*)(SparseDistributedRepresentation*)> callbacks;
-    void do_callbacks() {
-        for(auto func_ptr : callbacks)
-            func_ptr(this);
+    /**
+     * Remove the value from this SDR by clearing all of the valid flags.  Does
+     * not actually change any of the data.  Attempting to get the SDR's value
+     * immediately after this operation would raise an exception.
+     */
+    void clear() {
+        dense_valid      = false;
+        flatSparse_valid = false;
+        sparse_valid     = false;
     };
 
 public:
+    /**
+     * Use this method only in conjuction with sdr.load().
+     */
+    SparseDistributedRepresentation() {};
 
     /**
-     * Create an SDR object.  Initially this SDR has no value set.
+     * Create an SDR object.  Initially SDRs value is all zeros.
      *
      * @param dimensions A list of dimension sizes, defining the shape of the
      * SDR.
      */
-    SparseDistributedRepresentation( const vector<UInt> dimensions = vector<UInt>(0) ) {
+    SparseDistributedRepresentation( const vector<UInt> dimensions ) {
         dimensions_ = dimensions;
         // Calculate the SDR's size.
-        if( dimensions.size() ) {
-            size_ = 1;
-            for(UInt dim : dimensions)
-                size_ *= dim;
-        }
-        else {
-            size_ = 0;
+        size_ = 1;
+        for(UInt dim : dimensions)
+            size_ *= dim;
+        if ( dimensions.size() == 0 or size == 0 ) {
+            throw logic_error("SDR size is zero!");
```

Comment:
you can use `NTA_CHECK(size == 0) << "SDR size is zero!"` or NTA_ASSERT for this